types/wlr_gamma_control_v1.c

```diff
@@ -0,0 +1,265 @@
+		struct wl_resource *manager_resource, uint32_t id,
+		struct wl_resource *output_resource) {
+	struct wlr_gamma_control_manager_v1 *manager =
+		gamma_control_manager_from_resource(manager_resource);
+	struct wlr_output *output = wlr_output_from_resource(output_resource);
+
+	struct wlr_gamma_control_v1 *gamma_control =
+		calloc(1, sizeof(struct wlr_gamma_control_v1));
+	if (gamma_control == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	gamma_control->output = output;
+
+	uint32_t version = wl_resource_get_version(manager_resource);
+	gamma_control->resource = wl_resource_create(client,
+		&zwlr_gamma_control_v1_interface, version, id);
+	if (gamma_control->resource == NULL) {
+		free(gamma_control);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(gamma_control->resource, &gamma_control_impl,
+		gamma_control, gamma_control_handle_resource_destroy);
+
+	wl_signal_add(&output->events.destroy,
+		&gamma_control->output_destroy_listener);
+	gamma_control->output_destroy_listener.notify =
+		gamma_control_handle_output_destroy;
+
+	wl_list_init(&gamma_control->link);
+
+	if (!output->impl->set_gamma) {
+		zwlr_gamma_control_v1_send_failed(gamma_control->resource);
+		gamma_control_destroy(gamma_control);
+		return;
+	}
+
+	struct wlr_gamma_control_v1 *gc;
+	wl_list_for_each(gc, &manager->controls, link) {
+		if (gc->output == output) {
+			zwlr_gamma_control_v1_send_failed(gc->resource);
+			gamma_control_destroy(gc);
+			return;
+		}
+	}
+
+	wl_list_remove(&gamma_control->link);
+	wl_list_insert(&manager->controls, &gamma_control->link);
+	zwlr_gamma_control_v1_send_gamma_size(gamma_control->resource,
+		wlr_output_get_gamma_size(output));
+}
+
+static const struct zwlr_gamma_control_manager_v1_interface
+		gamma_control_manager_impl = {
+	.get_gamma_control = gamma_control_manager_get_gamma_control,
+};
+
+static void gamma_control_manager_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void gamma_control_manager_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_gamma_control_manager_v1 *manager = data;
+
+	struct wl_resource *resource = wl_resource_create(client,
+		&zwlr_gamma_control_manager_v1_interface, version, id);
+	if (resource == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(resource, &gamma_control_manager_impl,
+		manager, gamma_control_manager_handle_resource_destroy);
+	wl_list_insert(&manager->resources, wl_resource_get_link(resource));
+}
+
+void wlr_gamma_control_manager_v1_destroy(
+		struct wlr_gamma_control_manager_v1 *manager) {
+	if (!manager) {
+		return;
+	}
+	wl_list_remove(&manager->display_destroy.link);
+	struct wlr_gamma_control_v1 *gamma_control, *tmp;
+	wl_list_for_each_safe(gamma_control, tmp, &manager->controls, link) {
+		wl_resource_destroy(gamma_control->resource);
```

Comment:
Does this destroy the `gamma_control`s the client's still keeping a handle on?

Reply:
Yes. See https://github.com/swaywm/wlroots/issues/1096#issuecomment-408670983